jumanji/environments/routing/connector/generator.py

```diff
@@ -106,3 +125,462 @@ def __call__(self, key: chex.PRNGKey) -> State:
+
+    This generator places start positions randomly on the grid and performs a random walk from each.
+    Targets are placed at their terminuses.
+    """
+
+    def __init__(self, grid_size: int, num_agents: int) -> None:
+        """Instantiates a `RandomWalkGenerator.
+
+        Args:
+            grid_size: size of the square grid to generate.
+            num_agents: number of agents/paths on the grid.
+        """
+        super().__init__(grid_size, num_agents)
+
+    def __call__(self, key: chex.PRNGKey) -> State:
+        """Generates a `Connector` state that contains the grid and the agents' layout.
+
+        Args:
+            key: used to randomly generate the connector grid.
+
+        Returns:
+            A `Connector` state.
+        """
+        key, board_key = jax.random.split(key)
+        solved_grid, agents, grid = self.generate_board(board_key)
+        step_count = jnp.array(0, jnp.int32)
+
+        return State(key=key, grid=grid, step_count=step_count, agents=agents)
+
+    def generate_board(self, key: chex.PRNGKey) -> Tuple[chex.Array, Agent, chex.Array]:
+        """Generates solvable board using random walk.
+
+        Args:
+            key: random key.
+
+        Returns:
+            Tuple containing solved board, the agents and an empty training board.
+        """
+        grid = jnp.zeros((self.grid_size, self.grid_size), dtype=jnp.int32)
+        key, step_key = jax.random.split(key)
+        grid, agents = self._initialize_agents(key, grid)
+
+        stepping_tuple = (step_key, grid, agents)
+
+        _, grid, agents = jax.lax.while_loop(
+            self._continue_stepping, self._step, stepping_tuple
+        )
+
+        # Convert heads and targets to format accepted by generator
+        heads = agents.start.T
+        targets = agents.position.T
+
+        solved_grid = self.update_solved_board_with_head_target_encodings(
+            grid, tuple(heads), tuple(targets)
+        )
+        # Update agent information to include targets and positions after first step
+        agents.target = agents.position
+        agents.position = agents.start
+
+        agent_position_values = get_position(jnp.arange(self.num_agents))
+        agent_target_values = get_target(jnp.arange(self.num_agents))
+        # Populate an empty grid with heads and targets
+        grid = jnp.zeros((self.grid_size, self.grid_size), dtype=jnp.int32)
+        grid = grid.at[tuple(agents.start.T)].set(agent_position_values)
+        grid = grid.at[tuple(agents.target.T)].set(agent_target_values)
+        return solved_grid, agents, grid
+
+    def _step(
+        self, stepping_tuple: Tuple[chex.PRNGKey, chex.Array, Agent]
+    ) -> Tuple[chex.PRNGKey, chex.Array, Agent]:
+        """Takes one step for all agents."""
+        key, grid, agents = stepping_tuple
+        key, next_key = jax.random.split(key)
+        agents, grid = self._step_agents(key, grid, agents)
+        return next_key, grid, agents
+
+    def _step_agents(
```

Comment:
Hi - yes of course, I'll look into it - do you have a preferred solution? @sash-a @clement-bonnet 
My first thought would be to try instantiating the environment so I don't need to make other changes within the connector module

Reply:
I think instantiating the env might cause circular import issues and the generator shouldn't really know about the env, just it's state. So I think the better method would be to pull the 3 methods out into the utils file and then you can use those methods here and in the connector env. 

I think you don't need to worry about making changes to the connector module, as long as the tests still pass it should be fine :+1:

Reply:
I agree with @sash-a we should avoid code duplication. And the best solution for this is to extract some stepping code from the env to use in both the env and the generator.
However, since we are running out of time, I would lean towards accepting this solution for now and creating an issue to refactor this later.
Would this work for you two? @mwolinska @sash-a 
@mwolinska, would you mind making an issue for this, please?

Reply:
Yes works for me :+1:

Reply:
yes perfect, I'll do that now.